writer.go

```diff
@@ -585,15 +611,30 @@ func (w *writer) run() {
 
 	for !done {
 		var mustFlush bool
-
+		// lstMsg gets set when the next message would put the maxMessageBytes  over the limit.
+		// If a lstMsg exists we need to add it to the batch so we don't lose it.
+		if lstMsg != nil {
+			batch = append(batch, lstMsg.msg)
+			resch = append(resch, lstMsg.res)
+			batchSizeBytes += int(lstMsg.msg.message().size())
+			lstMsg = nil
+		}
 		select {
 		case wm, ok := <-w.msgs:
 			if !ok {
 				done, mustFlush = true, true
 			} else {
+				if int(wm.msg.message().size())+batchSizeBytes > w.maxMessageBytes {
```

Comment:
This is out of scope, but do you happen to know how `maxMessageBytes` relates to compression?  Since this calculation is happening prior to compression, I'm wondering whether there's an opportunity to get better compression ratios by continuing to add to the batch until the *compressed* payload is nearing `maxMessageBytes`.

Reply:
Yeah, this is a great question. 
All the batching is done before compression so it doens't related. So we are erring on the side of being conservative with the batch size.  
There is an opportunity here to get better compression, but we would need to do some re-working on where do the compression as currently it's done in the `write` and done to the entire batch. 

I think this is something to fix, but I would want to get this PR done first before getting after that.  

related: https://issues.apache.org/jira/browse/KAFKA-4169

Reply:
Cool.  Thanks for digging that Kafka ticket up.  I agree that it's a much bigger change that's out of scope for this PR.